File: python/src/base_human/checksum.py
"""Version 1 rolling polynomial checksum, spec section 6.2."""

from __future__ import annotations

from .basen import alphabet_index, encode_base_n
from .errors import INVALID_CHARACTER, HrcError
from .profile import PreparedProfile

_INITIAL_STATE = 17
_MULTIPLIER = 37


def checksum_value(profile: PreparedProfile, body: str, body_index: dict) -> int:
    """Return the checksum value in [0, modulus)."""
    modulus = profile.checksum_modulus
    state = _INITIAL_STATE
    for byte in profile.profile_id.encode("ascii"):
        state = (state * _MULTIPLIER + byte + 1) % modulus
    state = (state * _MULTIPLIER) % modulus
    for pos, ch in enumerate(body):
        value = body_index.get(ch)
        if value is None:
            raise HrcError(
                INVALID_CHARACTER, f"Symbol {ch!r} is not in the body alphabet"
            )
        state = (state * _MULTIPLIER + value + pos + 1) % modulus
    return state


def calculate_checksum(profile: PreparedProfile, body: str) -> str:
    """Compute the expected checksum string for a normalized body."""
    if profile.checksum_length == 0:
        return ""
    index = alphabet_index(profile.body_alphabet_norm)
    value = checksum_value(profile, body, index)
    return encode_base_n(value, profile.checksum_alphabet_norm, profile.checksum_length)

File: ruby/lib/base_human/checksum.rb
# frozen_string_literal: true

module BaseHuman
  # Version 1 checksum, spec section 6.2. Rolling polynomial over symbol
  # values, then modulus conversion into the checksum alphabet.
  module Checksum
    INITIAL_STATE = 17
    MULTIPLIER = 37

    module_function

    # Returns the checksum value in [0, modulus).
    def checksum_value(prepared, body, body_index = nil)
      body_index ||= BaseN.alphabet_index(prepared.body_alphabet)
      modulus = prepared.checksum_modulus

      state = INITIAL_STATE
      prepared.profile_id.each_byte do |byte|
        state = (state * MULTIPLIER + byte + 1) % modulus
      end
      state = (state * MULTIPLIER) % modulus
      body.each_char.each_with_index do |ch, pos|
        symbol_value = body_index[ch]
        unless symbol_value
          raise BasehError.new(
            "INVALID_CHARACTER",
            "Body symbol #{ch.inspect} is not in the body alphabet"
          )
        end
        state = (state * MULTIPLIER + symbol_value + pos + 1) % modulus
      end
      state
    end

    # Expected checksum string for a normalized body.
    def calculate_checksum(prepared, body, body_index = nil)
      return "" if prepared.checksum_length.zero?

      body_index ||= BaseN.alphabet_index(prepared.body_alphabet)
      value = checksum_value(prepared, body, body_index)
      BaseN.encode_base_n(value, prepared.checksum_alphabet, prepared.checksum_length)
    end
  end
end
